cmd/podman/images/scp.go

```diff
@@ -0,0 +1,410 @@
+		Example:           `podman image scp myimage:latest otherhost::`,
+	}
+)
+
+var (
+	scpOpts entities.ImageScpOptions
+)
+
+func init() {
+	registry.Commands = append(registry.Commands, registry.CliCommand{
+		Command: imageScpCommand,
+		Parent:  imageCmd,
+	})
+	scpFlags(imageScpCommand)
+}
+
+func scpFlags(cmd *cobra.Command) {
+	flags := cmd.Flags()
+	formatFlagName := "format"
+	flags.StringVar(&scpOpts.Save.Format, formatFlagName, define.V2s2Archive, "Save image to oci-archive, docker-archive, docker-dir")
+	_ = cmd.RegisterFlagCompletionFunc(formatFlagName, common.AutocompleteImageSaveFormat)
+	flags.BoolVarP(&scpOpts.Save.Quiet, "quiet", "q", false, "Suppress the output")
+}
+
+func scp(cmd *cobra.Command, args []string) (finalErr error) {
+	var (
+		tags []string
+		err  error
+	)
+	if scpOpts.Save.Quiet { // set quiet for both load and save
+		scpOpts.Load.Quiet = true
+	}
+	//	scpOpts.ToFrom = true                 // default to loading to the ssh client
+	f, err := ioutil.TempFile("", "temp") // open temp dir for load/save output
+	if err != nil {
+		return err
+	}
+	defer os.Remove(f.Name())
+	scpOpts.Save.Output = f.Name()
+	scpOpts.Load.Input = scpOpts.Save.Output
+	if err := parse.ValidateFileName(saveOpts.Output); err != nil {
+		return err
+	}
+	confR, err := config.NewConfig("") // create a hand made config for the remote engine since we might use remote and native at once
+	if err != nil {
+		return errors.Wrapf(err, "could not make config")
+	}
+	abiEng, err := registry.NewImageEngine(cmd, args) // abi native engine
+	if err != nil {
+		return errors.Wrapf(err, "could not make abi engine")
+	}
+
+	cfg, err := config.ReadCustomConfig() // get ready to set ssh destination if necessary
+	if err != nil {
+		return err
+	}
+	urls, iden, serv := parseLocation(args, cfg) // parses connection data and "which way" we are loading and saving
+
+	optionsTunnel := entities.PodmanConfig{Config: confR, EngineMode: entities.TunnelMode, Remote: true, URI: urls[0]}
+	if iden[0] != "" {
+		optionsTunnel = entities.PodmanConfig{Config: confR, EngineMode: entities.TunnelMode, Remote: true, Identity: iden[0], URI: urls[0]}
+	}
+
+	uri := []*urlP.URL{}
+	for i, url := range urls {
+		urlt, err := urlP.Parse(url) // create an actual url to pass to exec command
+		if err != nil {
+			return err
+		}
+		uri = append(uri, urlt)
+		if uri[i].User.Username() == "" {
+			if uri[i].User, err = getUserInfo(uri[i]); err != nil {
+				return err
+			}
+		}
+	}
+
+	confR.Engine = config.EngineConfig{Remote: true, CgroupManager: "cgroupfs", ServiceDestinations: serv} // pass the service dest (either remote or something else) to engine
+
+	rem := cmd.Root().Flags().Lookup("remote").Value // if remote was specified globally...
+	remS := rem.String()
+
+	switch {
+	case scpOpts.FromRemote: // if we want to load FROM the remote
+		_, err := execCommand(iden[0], uri[0], scpOpts.Save.Output, scpOpts.ImageName, args[0], false)
```

Comment:
(Maybe have the parser structure `args[…]` to return values, or values in `scpOpts` as well? Not at all sure.)

Reply:
See the discussion around `ImageName` — it seems to me that `parseLocation`, probably renamed to `parseArgs`, should be the only code reading `args`, and it should return (or write to `scpOpts`) well-named values.